File: 108.php
<?php

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution {

    /**
     * @param Integer[] $nums
     * @return TreeNode
     */
    function sortedArrayToBST($nums) {
        return $this->insert($nums, 0, count($nums) - 1);
    }

    function insert(array $nums, int $left, int $right) {
        if ($left > $right) {
            return null;
        }

        $middle = (int)(($left + $right) / 2);
        $node = new TreeNode($nums[$middle]);

        $node->left = $this->insert($nums, $left, $middle - 1);
        $node->right = $this->insert($nums, $middle + 1, $right);

        return $node;
    }

}

File: go/array/108.go
package main


func sortedArrayToBST(nums []int) *TreeNode {
	length := len(nums)
	if length == 0 {
		return nil
	}
	if length == 1 {
		return &TreeNode{Val: nums[0]}
	}
	middle := (length - 1) / 2
	left := nums[:middle]
	right := nums[(middle+1):]

	node := new(TreeNode)
	node.Val = nums[middle]
	node.Left = sortedArrayToBST(left)
	node.Right = sortedArrayToBST(right)

	return node
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}
